cssp.c

```diff
@@ -115,18 +115,20 @@ cssp_gss_mech_available(gss_OID mech)
 }
 
 static RD_BOOL
-cssp_gss_get_service_name(char *server, gss_name_t * name)
+cssp_gss_get_service_name(struct addrinfo *ai, gss_name_t * name)
 {
 	gss_buffer_desc output;
 	OM_uint32 major_status, minor_status;
-
+	char fqdn[64];
+	int size;
 	const char service_name[] = "TERMSRV";
-
 	gss_OID type = (gss_OID) GSS_C_NT_HOSTBASED_SERVICE;
-	int size = (strlen(service_name) + 1 + strlen(server) + 1);
+
+	getnameinfo(ai->ai_addr, ai->ai_addrlen, fqdn, sizeof(fqdn), NULL, 0, 0);
+	size = (strlen(service_name) + 1 + strlen(fqdn) + 1);
 
 	output.value = malloc(size);
-	snprintf(output.value, size, "%s@%s", service_name, server);
+	snprintf(output.value, size, "%s@%s", service_name, fqdn);
```

Comment:
Won't this break systems where you don't have forward and reverse lookups in sync? `AI_CANONNAME` is probably a better idea, possibly with `getnameinfo()` as a fallback for ip addresses (if `AI_CANONNAME' doesn't do reverse lookup for you).

Reply:
As I understand and have verified, AI_CANONNAME would be the resolved name for the passed hostname to getaddrinfo(), this name can be a RDS farm name with round robin and this is not allowed to be used with a SPN which needs to be the actual hostname FQDN. So we can't use that.

Reply:
Another problem exists, it resolves the first item in list and assume that this is the host connected to. Should be the entry int the list that a successful connection was made to. In most cases this will work, except when the first entry from a round robin is down and second one is used for connection.

Reply:
Shouldn't you add the farm SPN to each server in that case? I can find some people referring to using `setspn`, but this Microsoft blog seem to indicate that the connection broken can keep track of a common Kerberos identity:

https://cloudblogs.microsoft.com/enterprisemobility/2009/05/20/creating-kerberos-identity-for-rd-session-host-farms-part-i-using-the-remote-desktop-services-provider-for-windows-powershell/

Reply:
That was a feature for only Windows 2008 that is deprecated. See first response to that blog post. When deploying a RDS session host a SPN is automatically created for the host which makes our approach to work out of the box against both 2008 and later.

Reply:
Alright. And no issues with the security? Doing a reverse lookup for normal Kerberos makes it vulnerable to spoofing, so the same might be true here?

`AI_CANONNAME` seems to have issues as well, though: :/
http://k5wiki.kerberos.org/wiki/Projects/Trust_KDC-local_name_resolution
http://mailman.mit.edu/pipermail/kerberos/2011-July/017313.html